File: providers/openai.py
from __future__ import annotations

import asyncio
import json
import logging

import openai as oai

from ..config import Config
from .base import Provider, StreamResponse, ToolUse

log = logging.getLogger(__name__)

_MAX_RETRIES = 6


class OpenAIProvider(Provider):
    def __init__(self, config: Config):
        kwargs: dict = {"api_key": config.api_key or None}
        if config.base_url:
            kwargs["base_url"] = config.base_url
        self._client = oai.AsyncOpenAI(**kwargs)
        self._model = config.model

    async def stream(self, messages: list[dict], tools: list[dict], model: str) -> StreamResponse:
        delay = 1.0
        for attempt in range(_MAX_RETRIES):
            try:
                return await self._call(messages, tools, model)
            except oai.RateLimitError:
                if attempt == _MAX_RETRIES - 1:
                    raise
                await asyncio.sleep(delay)
                delay = min(delay * 2, 60)

    @staticmethod
    def _convert_messages(messages: list[dict]) -> list[dict]:
        """Convert Anthropic-style content blocks to OpenAI format."""
        out = []
        for msg in messages:
            role = msg["role"]
            content = msg.get("content", "")
            if isinstance(content, str):
                out.append({"role": role, "content": content})
                continue
            # Anthropic list content
            if role == "assistant":
                text_parts = [b["text"] for b in content if b.get("type") == "text"]
                tool_calls = [
                    {"id": b["id"], "type": "function", "function": {
                        "name": b["name"],
                        "arguments": json.dumps(b["input"]),
                    }}
                    for b in content if b.get("type") == "tool_use"
                ]
                entry: dict = {"role": "assistant", "content": " ".join(text_parts) or None}
                if tool_calls:
                    entry["tool_calls"] = tool_calls
                out.append(entry)
            elif role == "user":
                # may be tool_result blocks
                tool_results = [b for b in content if b.get("type") == "tool_result"]
                if tool_results:
                    for b in tool_results:
                        out.append({
                            "role": "tool",
                            "tool_call_id": b["tool_use_id"],
                            "content": b["content"] if isinstance(b["content"], str)
                                       else json.dumps(b["content"]),
                        })
                else:
                    text = " ".join(b.get("text", "") for b in content if b.get("type") == "text")
                    out.append({"role": "user", "content": text})
            else:
                out.append(msg)
        return out

    async def _call(self, messages: list[dict], tools: list[dict], model: str) -> StreamResponse:
        kwargs: dict = dict(model=model, messages=self._convert_messages(messages))
        if tools:
            # convert Anthropic tool schema to OpenAI function schema
            kwargs["tools"] = [
                {"type": "function", "function": {
                    "name": t["name"],
                    "description": t.get("description", ""),
                    "parameters": t.get("input_schema", {}),
                }}
                for t in tools
            ]
            kwargs["tool_choice"] = "auto"

        resp = await self._client.chat.completions.create(**kwargs)
        choice = resp.choices[0]
        msg = choice.message

        text = msg.content or ""
        tool_uses: list[ToolUse] = []
        if msg.tool_calls:
            for tc in msg.tool_calls:
                tool_uses.append(ToolUse(
                    id=tc.id,
                    name=tc.function.name,
                    input=json.loads(tc.function.arguments),
                ))

        stop = "tool_use" if tool_uses else "end_turn"
        if choice.finish_reason == "length":
            stop = "max_tokens"

        usage = resp.usage
        return StreamResponse(
            text=text,
            stop_reason=stop,
            tool_uses=tool_uses,
            input_tokens=usage.prompt_tokens if usage else 0,
            output_tokens=usage.completion_tokens if usage else 0,
        )

File: src/providers/openai.ts
import OpenAI from "openai";
import type { Config } from "../config.ts";
import { Provider, makeStreamResponse } from "./base.ts";
import type { Message, ApiTool, StreamResponse, ToolUse, ContentBlock } from "./base.ts";

const MAX_RETRIES = 5;

export class OpenAIProvider extends Provider {
  private readonly client: OpenAI;

  constructor(config: Config) {
    super();
    const opts: ConstructorParameters<typeof OpenAI>[0] = {
      apiKey: config.apiKey || undefined,
    };
    if (config.baseUrl) {
      opts.baseURL = config.baseUrl;
    }
    this.client = new OpenAI(opts);
  }

  async stream(messages: Message[], tools: ApiTool[], model: string): Promise<StreamResponse> {
    let delay = 1.0;
    for (let attempt = 0; attempt < MAX_RETRIES; attempt++) {
      try {
        return await this._call(messages, tools, model);
      } catch (err) {
        if (err instanceof OpenAI.RateLimitError) {
          if (attempt === MAX_RETRIES - 1) throw err;
          await sleep(delay * 1000);
          delay = Math.min(delay * 2, 60);
          continue;
        }
        throw err;
      }
    }
    throw new Error("unreachable");
  }

  private static convertMessages(messages: Message[]): OpenAI.ChatCompletionMessageParam[] {
    const out: OpenAI.ChatCompletionMessageParam[] = [];

    for (const msg of messages) {
      const { role, content } = msg;

      if (typeof content === "string") {
        out.push({ role: role as "user" | "assistant" | "system", content });
        continue;
      }

      // list content blocks
      if (role === "assistant") {
        const textParts = (content as ContentBlock[])
          .filter((b) => b.type === "text")
          .map((b) => (b as { type: "text"; text: string }).text)
          .join(" ");

        const toolCalls = (content as ContentBlock[])
          .filter((b) => b.type === "tool_use")
          .map((b) => {
            const tu = b as { type: "tool_use"; id: string; name: string; input: Record<string, unknown> };
            return {
              id: tu.id,
              type: "function" as const,
              function: {
                name: tu.name,
                arguments: JSON.stringify(tu.input),
              },
            };
          });

        const entry: OpenAI.ChatCompletionAssistantMessageParam = {
          role: "assistant",
          content: textParts || null,
        };
        if (toolCalls.length > 0) {
          entry.tool_calls = toolCalls;
        }
        out.push(entry);
      } else if (role === "user") {
        const toolResults = (content as ContentBlock[]).filter((b) => b.type === "tool_result");
        if (toolResults.length > 0) {
          for (const b of toolResults) {
            const tr = b as { type: "tool_result"; tool_use_id: string; content: string };
            out.push({
              role: "tool",
              tool_call_id: tr.tool_use_id,
              content: typeof tr.content === "string" ? tr.content : JSON.stringify(tr.content),
            });
          }
        } else {
          const text = (content as ContentBlock[])
            .filter((b) => b.type === "text")
            .map((b) => (b as { type: "text"; text: string }).text)
            .join(" ");
          out.push({ role: "user", content: text });
        }
      } else {
        out.push({ role: role as "system", content: typeof content === "string" ? content : JSON.stringify(content) });
      }
    }

    return out;
  }

  private async _call(messages: Message[], tools: ApiTool[], model: string): Promise<StreamResponse> {
    const params: OpenAI.ChatCompletionCreateParams = {
      model,
      messages: OpenAIProvider.convertMessages(messages),
      stream: false,
    };

    if (tools.length > 0) {
      params.tools = tools.map((t) => ({
        type: "function" as const,
        function: {
          name: t.name,
          description: t.description ?? "",
          parameters: t.input_schema as Record<string, unknown>,
        },
      }));
      params.tool_choice = "auto";
    }

    const resp = await this.client.chat.completions.create(params);
    const choice = resp.choices[0]!;
    const msg = choice.message;

    const text = msg.content ?? "";
    const toolUses: ToolUse[] = [];

    if (msg.tool_calls) {
      for (const tc of msg.tool_calls) {
        toolUses.push({
          id: tc.id,
          name: tc.function.name,
          input: JSON.parse(tc.function.arguments) as Record<string, unknown>,
        });
      }
    }

    let stopReason: "end_turn" | "tool_use" | "max_tokens" = "end_turn";
    if (toolUses.length > 0) stopReason = "tool_use";
    if (choice.finish_reason === "length") stopReason = "max_tokens";

    const usage = resp.usage;
    return makeStreamResponse(
      text,
      stopReason,
      toolUses,
      usage?.prompt_tokens ?? 0,
      usage?.completion_tokens ?? 0,
    );
  }
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}
